### PR TITLE
Serialize entire struct to/from guestinfo

### LINKED ISSUE #402: Serialize entire struct to/from guestinfo
Write an Encoder/Decoder for guestinfo storage of structures.
We wish to be able to serialize structs with:
maps
embedded types
the exec.Cmd struct

Serialization should be able to be sparse, with only values that are non-default being serialized.
Deserialization should be able to update an existing object

The intent here is that we be able to:
1. populate an exec.Cmd object with the command details
2. serialize the structure
3. deserialize the structure
4. launch the command
5. wait for it to complete
6. serialize the structure, including exit code
7. deserialize structure and retrieve exit code

Do this by adding annotations to the struct with the explicit extraConfig path at which to store the value - this will be replaced eventually by calculated paths derived from other annotations, however it's a closer match than making everything one giant blob.

### PR DESCRIPTION
**Pull request by caglar10ur**
Introduce the encoder. It encodes "vic" tagged structs and their members
into []types.BaseOptionValue so that VM spec can directly consume them.

Following is the result of calling encoder on T
```
    T := struct {
        ExecutorConfig ExecutorConfig `vic:"0.1" scope:"hidden" key:"executorconfig"`
    }{
        ExecutorConfig{
            Sessions: map[string]SessionConfig{
                "Session1": SessionConfig{
                    Common: Common{
                        ID:   "SessionID",
                        Name: "SessionName",
                    },
                    Tty: true,
                    Cmd: Cmd{
                        Path: "/vmware",
                        Args: []string{"/bin/imagec", "-standalone"},
                        Env:  []string{"PATH=/bin", "USER=imagec"},
                        Dir:  "/",
                    },
                },
            },
        },
    }

    encode_test.go:327: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"guestinfo/executorconfig/common~id", Value:""}
    encode_test.go:327: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"guestinfo/executorconfig/common~name", Value:""}
    encode_test.go:327: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"guestinfo/executorconfig/common~notes", Value:""}
    encode_test.go:327: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"executorconfig~sessions|Session1~common~id", Value:"SessionID"}
    encode_test.go:327: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"executorconfig~sessions|Session1~common~name", Value:"SessionName"}
    encode_test.go:327: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"executorconfig~sessions|Session1~common~notes", Value:""}
    encode_test.go:327: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"executorconfig~sessions|Session1~cmd~path", Value:"/vmware"}
    encode_test.go:327: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"executorconfig~sessions|Session1~cmd~args", Value:"/bin/imagec|-standalone"}
    encode_test.go:327: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"executorconfig~sessions|Session1~cmd~env", Value:"PATH=/bin|USER=imagec"}
    encode_test.go:327: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"executorconfig~sessions|Session1~cmd~dir", Value:"/"}
    encode_test.go:327: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"executorconfig~sessions|Session1~tty", Value:"true"}
    encode_test.go:327: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"executorconfig~sessions", Value:"Session1"}
```
Map tests are disabled because of github.com/stretchr/testify/issues/288

### CODE REVIEW

**Comment by caglar10ur** [comment]
This is here as a draft. Encoder needs to handle corner cases, tests needs to be improved and some documentation will hopefully follow in the near future.

**Comment by emlin** [comment]
So, I didn't see decode, will that be in another pull request?

File: pkg/vsphere/guestinfo/encode.go [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return strconv.FormatInt(field.Int(), 10)
+	case reflect.Bool:
+		return strconv.FormatBool(field.Bool())
+	case reflect.String:
+		return field.String()
+	case reflect.Float32, reflect.Float64:
+		return strconv.FormatFloat(field.Float(), 'E', -1, 64)
+	default:
+		return ""
```
Comment (emlin) [inline pkg/vsphere/guestinfo/encode.go]: what about other types, e.g. time?

**Comment by caglar10ur** [comment]
Yep, decode will come later as I'm still working on it.

File: pkg/vsphere/guestinfo/encode.go [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+	// Invalid value
+	Invalid = 1 << iota
+	// Hidden value
+	Hidden
+	// ReadOnly value
+	ReadOnly
+	// ReadWrite value
+	ReadWrite
+	// NonPersistent value
+	NonPersistent
```
Comment (fdawg4l) [inline pkg/vsphere/guestinfo/encode.go]: `NonPersistent` is just ` not Persistent`.  Can we remove the `Non` modifier and just make this `Persistent`?

File: pkg/vsphere/guestinfo/encode.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+	// Volatile value
+	Volatile
```
Comment (fdawg4l) [inline pkg/vsphere/guestinfo/encode.go]: What does volatile mean?  Does it mean the value can be updated outside of read-modify-write (like the `volatile` keyword)?

File: pkg/vsphere/guestinfo/encode.go (hunk 2 above)
Comment (caglar10ur) [inline pkg/vsphere/guestinfo/encode.go]: Those values are coming from @hickeng's docs. I'm fine changing it to anything as long as he is OK :)

File: pkg/vsphere/guestinfo/encode.go (hunk 3 above)
Comment (caglar10ur) [inline pkg/vsphere/guestinfo/encode.go]: https://github.com/hickeng/vic/blob/467/doc/design/configuration.md

Comment (hickeng) [inline pkg/vsphere/guestinfo/encode.go]: I'd be happy dropping volatile for now - it's in there as a speculative thing and adds nothing in this initial encode/decode pass. I we get to a full 'managed object' (and I know that's going to lead to questions), then it becomes useful.

File: pkg/vsphere/guestinfo/encode.go (hunk 2 above)
Comment (hickeng) [inline pkg/vsphere/guestinfo/encode.go]: NonPersistent is because it's the rare case where Persistent is the general case. If we use nonpersistent as the bool it can be omitted from the tag set for most fields. This is for conciseness and readability.

The reason I don't suggest the same thing for read-only/read-write is that people tend to assume that a value is read-write, yet that's not our common case. Being explicit as to which for that pairing should improve readability.

File: pkg/vsphere/guestinfo/encode.go (hunk 1 above)
Comment (hickeng) [inline pkg/vsphere/guestinfo/encode.go]: We could add time - I'm not sure what configuration element requires it currently unless creation/poweron time isn't part of the standard vmodl. I'd suggest adding things only as we need them for now.

File: pkg/vsphere/guestinfo/encode.go [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+	var guestinfo string
+	var seperator string
+
+	// Trim the whitespaces
+	key = strings.TrimSpace(key)
+
+	if scope&Hidden != 0 {
+		guestinfo = ""
+		seperator = "~"
```
Comment (hickeng) [inline pkg/vsphere/guestinfo/encode.go]: I suggest using / here as well - it's a fail safe in terms of read-only and I don't know what the underlying guestinfo mechanism does if you don't use / or . separators.

File: pkg/vsphere/guestinfo/encode.go [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+				config = append(config, EncodeWithPrefix(value.Index(i).Interface(), key)...)
+			} else {
+				values = append(values, ToString(value.Index(i)))
+			}
+		}
+		// set the slice key with the values seperated by |
+		if len(values) > 0 {
+			// sort the values before joining
+			sort.Strings(values)
+			// prefix~ contains the items
```
Comment (hickeng) [inline pkg/vsphere/guestinfo/encode.go]: again, not sure what the impact of using ~ is on the underlying guestinfo mechanism. ditto for |

File: pkg/vsphere/guestinfo/encode.go [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+			keys := strings.Split(tags.Get("key"), ",")
+			key, properties := keys[0], keys[1:]
+
+			// returns the i'th field of the struct i
+			field := value.Field(i)
+
+			// process properties
+			if len(properties) > 0 {
+				// do not recurse into nested type
+				if properties[0] == "omitnested" {
```
Comment (hickeng) [inline pkg/vsphere/guestinfo/encode.go]: will need to doc that omitnested has to be first in the list.

File: pkg/vsphere/guestinfo/encode_test.go [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+}
+
+type SessionConfig struct {
+	Common `vic:"0.1" scope:"hidden" key:"common"`
+
+	Cmd Cmd `vic:"0.1" scope:"hidden" key:"cmd"`
+
+	Tty bool `vic:"0.1" scope:"hidden" key:"tty"`
+}
+
```
Comment (hickeng) [inline pkg/vsphere/guestinfo/encode_test.go]: recommend that we have a test that has some json tags for example - a case where there are tags that are not 'vic'. Don't bother about a mix of both vic and something else.

**Comment by hickeng** [comment]
LGTM - minor comments, but nothing that prevents merging in current state.

File: pkg/vsphere/guestinfo/encode.go (hunk 4 above)
Comment (caglar10ur) [inline pkg/vsphere/guestinfo/encode.go]: It seems to be working here, I'll see what I can

```
[vagrant@vagrant:/opt/go/src/github.com/vmware/vic(decode)] bin/rpctool -set "a|b~c" "test"
[vagrant@vagrant:/opt/go/src/github.com/vmware/vic(decode)] bin/rpctool -get "a|b~c"
test
```

File: pkg/vsphere/guestinfo/encode.go (hunk 5 above)
Comment (caglar10ur) [inline pkg/vsphere/guestinfo/encode.go]: ^

File: pkg/vsphere/guestinfo/encode.go (hunk 1 above)
Comment (caglar10ur) [inline pkg/vsphere/guestinfo/encode.go]: Will try to add it (it should be straightforward) as soon as I finish decode.

File: pkg/vsphere/guestinfo/encode.go (hunk 3 above)
Comment (caglar10ur) [inline pkg/vsphere/guestinfo/encode.go]: You were talking about exporting volatile objects via getter/setters, we can do that if we still want that. Or we can drop and add it back when we need them

File: pkg/vsphere/guestinfo/encode.go (hunk 6 above)
Comment (caglar10ur) [inline pkg/vsphere/guestinfo/encode.go]: Yep will try to write a short something to document the syntax, our assumptions and rules

File: pkg/vsphere/guestinfo/encode_test.go (hunk 7 above)
Comment (caglar10ur) [inline pkg/vsphere/guestinfo/encode_test.go]: Good point, will add them

**Comment by caglar10ur** [comment]
One big question I have at this point is the package name. Guestinfo is too generic and couldn't find any other. Any suggestions?

File: pkg/vsphere/guestinfo/encode.go (hunk 3 above)
Comment (hickeng) [inline pkg/vsphere/guestinfo/encode.go]: Add it when we need it - we don't need to remove volatile from the code, but doc that currently has no effect.

**Comment by hickeng** [comment]
pkg/vsphere/extraconfig

**Comment by caglar10ur** [comment]
Renamed to extraconfig, add time.Time support. Will merge after adding the json test cases and reading a brief readme

**Comment by caglar10ur** [comment]
Decoder piece landed as well so it would be really great to have one more pass on latest code

File: pkg/vsphere/extraconfig/doc.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+The implementation understands the following set of annotations and map the fields to appropriate extraConfig keys - in the case where the key describes a boolean state, ommitting the annotation implies the opposite:
+
+hidden - hidden from GuestOS
+read-only - value can only be modified via vSphere APIs
+read-write - value can be modified
+non-persistent - value will be lost on VM reboot
+volatile - field is not exported directly, but via a function that freshens the value each time)
+
+The struct fields are required to be annotated with the "vic" tag, otherwise extraconfig package simply skips them. Scope and key tags are also required.append
```
Comment (andrewtchin) [inline pkg/vsphere/extraconfig/doc.go]: Is the end supposed to be `.append`? If so, what is `required.append`? Same with line 18

Comment (caglar10ur) [inline pkg/vsphere/extraconfig/doc.go]: Done

**Comment by caglar10ur** [comment]
It's important to note that this version does not test error cases and not fully supports decoding into an already existing struct.

**Comment by emlin** [comment]
Awesome! Just wondering which kind of existing structure is not supported by the decoding?

**Comment by caglar10ur** [comment]
@emlin I didn't test every combination but I believe we can only merge/overwrite maps right now.

File: pkg/vsphere/extraconfig/decode.go [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+		if err != nil {
+			log.Errorf("Failed to convert: %s", err.Error())
+			return field
+		}
+		return reflect.ValueOf(s)
+
+	}
+	log.Debugf("Invalid Kind: %s", field.Kind())
+
+	return field
```
Comment (hickeng) [inline pkg/vsphere/extraconfig/decode.go]: should this be nil? Is field == nil for the field type?

File: pkg/vsphere/extraconfig/decode.go [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+}
+
+// decodeWithPrefix convert []types.BaseOptionValue to type dest
+func decodeWithPrefix(kv map[string]string, dest interface{}, prefix string) interface{} {
+	// value representing the run-time data
+	value := reflect.ValueOf(dest).Elem()
+	log.Debugf("Value: %#v", value)
```
Comment (hickeng) [inline pkg/vsphere/extraconfig/decode.go]: I've found utility in having a global var for 'verbose' - that gets set by tests but not during standard operation. workaround for missing debug levels.

File: pkg/vsphere/extraconfig/decode.go [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+// fromString converts string representation of a basic type to basic type
+func fromString(field reflect.Value, value string) reflect.Value {
+	switch field.Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		s, err := strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			log.Errorf("Failed to convert: %s", err.Error())
```
Comment (dougm) [inline pkg/vsphere/extraconfig/decode.go]: Would be good to provide more context with these errors, strconv doesn't provide much iirc.  Maybe include kind and value in the messages?

File: pkg/vsphere/extraconfig/decode.go [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+			log.Debugf("Keys: %#v Properties: %#v", key, properties)
+
+			// returns the i'th field of the struct i
+			field := value.Field(i)
+
+			// process properties
+			if len(properties) > 0 {
+				// do not recurse into nested type
+				if properties[0] == "omitnested" {
+					log.Debugf("Skipping %s (omitnested)", key)
```
Comment (hickeng) [inline pkg/vsphere/extraconfig/decode.go]: can we tell if there are values that would be unpacked under this type if not omitted? Can we issue a warning if that's the case?

File: Makefile [hunk 13]
```diff
@@ -166,7 +166,6 @@ govet: $(find . -type f -name '*.go' -not -path "./vendor/*" -not -path "apiserv
 	@echo checking go vet...
 	@$(GO) tool vet -all $$(find . -type f -name '*.go' -not -path "./vendor/*")
 	@$(GO) tool vet -shadow $$(find . -type f -name '*.go' -not -path "./vendor/*")
-	@touch $@
```
Comment (dougm) [inline Makefile]: wondered why that was there.  suppose you could revert b2c4261 with this gone, but no biggie.

File: pkg/vsphere/extraconfig/decode.go [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+			case reflect.Struct:
+				log.Debugf("Struct begin")
+
+				// type cast struct to supported types (eg; time.Time, url.Url etc.)
+				switch field.Interface().(type) {
+				case time.Time:
+					// from https://golang.org/src/time/format.go?s=12854:12883#L407
+					t, err := time.Parse("2006-01-02 15:04:05.999999999 -0700 MST", v)
+					if err != nil {
+						log.Errorf("Conversion failed: %s", err.Error())
```
Comment (dougm) [inline pkg/vsphere/extraconfig/decode.go]: maybe more context in the message here too.

**Comment by dougm** [comment]
LGTM

/giphy epic reflection

File: pkg/vsphere/extraconfig/extraconfig_test.go [hunk 15]
```diff
@@ -0,0 +1,584 @@
...
+		3.14,
+		"Grrr",
+	}
+
+	encoded := Encode(Struct)
+	expected := []types.BaseOptionValue{
+		&types.OptionValue{DynamicData: types.DynamicData{}, Key: "guestinfo.int", Value: "42"},
+		&types.OptionValue{DynamicData: types.DynamicData{}, Key: "guestinfo.bool", Value: "true"},
+		&types.OptionValue{DynamicData: types.DynamicData{}, Key: "guestinfo.float", Value: "3.14E+00"},
+		&types.OptionValue{DynamicData: types.DynamicData{}, Key: "guestinfo.string", Value: "Grrr"},
```
Comment (hickeng) [inline pkg/vsphere/extraconfig/extraconfig_test.go]: These are the only read-write keys I see - have you tried it with complex types where the outside type is read-only and a specific field is read-write?

File: pkg/vsphere/extraconfig/decode.go (hunk 9 above)
Comment (caglar10ur) [inline pkg/vsphere/extraconfig/decode.go]: To differentiate an unset value from a nil one ? If so I think we are not handling that at all right now. Do you think it's required in the first pass? I was planning to address that later

File: pkg/vsphere/extraconfig/decode.go (hunk 10 above)
Comment (caglar10ur) [inline pkg/vsphere/extraconfig/decode.go]: Having a global makes sense. Not sure whether we want tests to generate a lot of output though. I guess we can enable it if the test fails for post debugging.

File: pkg/vsphere/extraconfig/decode.go (hunk 11 above)
Comment (caglar10ur) [inline pkg/vsphere/extraconfig/decode.go]: Will do that

File: Makefile (hunk 13 above)
Comment (caglar10ur) [inline Makefile]: Yep, that's my plan :)

File: pkg/vsphere/extraconfig/decode.go (hunk 14 above)
Comment (caglar10ur) [inline pkg/vsphere/extraconfig/decode.go]: Done

File: pkg/vsphere/extraconfig/extraconfig_test.go (hunk 15 above)
Comment (caglar10ur) [inline pkg/vsphere/extraconfig/extraconfig_test.go]:
```
	type Inside struct {
		ID   string `vic:"0.1" scope:"read-write" key:"id"`
		Name string `vic:"0.1" scope:"read-write" key:"name"`
	}
	type Outside struct {
		Inside Inside `vic:"0.1" scope:"read-only" key:"inside"`
		ID     string `vic:"0.1" scope:"read-write" key:"id"`
		Name   string `vic:"0.1" scope:"read-write" key:"name"`
	}
```
gives 
```
	extraconfig_test.go:611: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"guestinfo.guestinfo/inside.id", Value:"inside"}
	extraconfig_test.go:611: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"guestinfo.guestinfo/inside.name", Value:"Inside"}
	extraconfig_test.go:611: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"guestinfo.id", Value:"outside"}
	extraconfig_test.go:611: &types.OptionValue{DynamicData:types.DynamicData{}, Key:"guestinfo.name", Value:"Outside"}
```
which looks OK to me. Will be adding that as a test case

Comment (hickeng) [inline pkg/vsphere/extraconfig/extraconfig_test.go]: There are multiple guestinfo's prefixed for the inside keys - I suspect that's an unintended consequence.
I am also unsure that `guestinfo.guestinfo/inside.id` is actually read/write. I think a / anywhere in the path makes it read only, but that's something we'll have to confirm experimentally.

Comment (caglar10ur) [inline pkg/vsphere/extraconfig/extraconfig_test.go]: Yeah multiple guestinfo looks like a bug. I'm not sure why we want id as read-write in this example, Outside marks the whole struct read-only

Comment (caglar10ur) [inline pkg/vsphere/extraconfig/extraconfig_test.go]: The multiple guestinfo prefix issue resolved with the latest revision.

Comment (hickeng) [inline pkg/vsphere/extraconfig/extraconfig_test.go]: @caglar10ur should probably be an error of some kind is it's tagged as read/write but cannot be made that way - we can wait and see how we want this to be resolved.

**Comment by hickeng** [comment]
@caglar10ur LGTM as first pass.
